src/Microsoft.OData.Core/ODataBatchJsonWriter.cs

```diff
@@ -0,0 +1,387 @@
+            // so use the URL resolver from the batch message instead.
+            this.CurrentOperationResponseMessage = ODataBatchOperationResponseMessage.CreateWriteMessage(
+                this.JsonLightOutputContext.OutputStream,
+                /*operationListener*/ this,
+                this.urlResolver.BatchMessageUrlResolver,
+                contentId);
+            this.SetState(BatchWriterState.OperationCreated);
+
+            Debug.Assert(this.currentOperationContentId == null, "The Content-ID header is only supported in request messages.");
+
+            // Start the Json object for the response
+            this.WriteStartBoundaryForOperation();
+
+            return this.CurrentOperationResponseMessage;
+        }
+
+        /// <summary>
+        /// Write any pending headers for the current operation message (if any).
+        /// </summary>
+        /// <param name="reportMessageCompleted">
+        /// A flag to control whether after writing the pending data we report writing the message to be completed or not.
+        /// </param>
+        internal override void WritePendingMessageData(bool reportMessageCompleted)
+        {
+            if (this.CurrentOperationMessage != null)
+            {
+                Debug.Assert(this.JsonLightOutputContext.JsonWriter != null, "Must have a batch writer if pending data exists.");
+
+                if (this.CurrentOperationResponseMessage != null)
+                {
+                    Debug.Assert(this.JsonLightOutputContext.WritingResponse, "If the response message is available we must be writing response.");
+
+                    this.jsonWriter.WriteName("id");
+                    this.jsonWriter.WriteValue(this.CurrentOperationResponseMessage.ContentId);
+
+                    this.jsonWriter.WriteName("status"); 
+                    this.jsonWriter.WriteValue(this.CurrentOperationResponseMessage.StatusCode);
+                }
+
+                // headers attribute
+                this.jsonWriter.WriteName("headers");
+                this.jsonWriter.StartObjectScope();
+                IEnumerable<KeyValuePair<string, string>> headers = this.CurrentOperationResponseMessage.Headers;
+                if (headers != null)
+                {
+                    foreach (KeyValuePair<string, string> headerPair in headers)
+                    {
+                        this.jsonWriter.WriteName(headerPair.Key);
+                        this.jsonWriter.WriteValue(headerPair.Value);
+                    }
+                }
+                this.jsonWriter.EndObjectScope();
+
+                
+                // body attribute
+                if (!ODataBatchWriterUtils.HasResponseBody(this.CurrentOperationResponseMessage))
+                {
+                    // Close the individual response object now since there are no content.
+                    this.jsonWriter.EndObjectScope();

```

Comment:
Where does this get closed if there is a body? Why doesn't it get closed the same place whether or not there is a body?